File: ext/RmlUi.Native/src/VariableDefinition.cpp
#include "RmlNative.h"
#include "RmlUi/Core/DataVariable.h"
#include <string>

typedef bool(*onGet)(void* ptr, Rml::Variant*& variant);
typedef bool(*onSet)(void* ptr, const Rml::Variant& variant);
typedef int(*onSize)(void* ptr);
typedef Rml::DataVariable*(*onChild)(void* ptr, const char* name, int offset);

class VariableDefinition : Rml::VariableDefinition {
private:
    ::onGet m_onGet;
    ::onSet m_onSet;
    ::onSize m_onSize;
    ::onChild m_onChild;

public:
    explicit VariableDefinition(Rml::DataVariableType type, ::onGet onGet, ::onSet onSet, ::onSize onSize, ::onChild onChild) : Rml::VariableDefinition(type) {
        m_onGet = onGet;
        m_onSet = onSet;
        m_onSize = onSize;
        m_onChild = onChild;
    }

    bool Get(void* ptr, Rml::Variant& variant) override {
      Rml::Variant* variant_ptr = &variant;
      auto res = (*m_onGet)(ptr, variant_ptr);
      variant = *variant_ptr;
      return res;
    }

    bool Set(void* ptr, const Rml::Variant& variant) override {
        return (*m_onSet)(ptr, variant);
    }

    int Size(void* ptr) override {
        return (*m_onSize)(ptr);
    }

    Rml::DataVariable Child(void* ptr, const Rml::DataAddressEntry& address) override {
        return *(*m_onChild)(ptr, address.name.c_str(), address.index);
    }
};

RMLUI_CAPI void *rml_VariableDefinition_New(Rml::DataVariableType type, ::onGet onGet, ::onSet onSet, ::onSize onSize, ::onChild onChild) {
    return new VariableDefinition(type, onGet, onSet, onSize, onChild);
}

RMLUI_CAPI void rml_VariableDefinition_Free(VariableDefinition* definition) {
    delete definition;
}

File: ext/RmlUI.Net/VariableDefinition.cs
using System;
using System.Text;

namespace RmlUiNet {
    public abstract class VariableDefinition : RmlBase<VariableDefinition> {
        private Native.VariableDefinition.OnGet _onGet;
        private Native.VariableDefinition.OnSet _onSet;
        private Native.VariableDefinition.OnSize _onSize;
        private Native.VariableDefinition.OnChild _onChild;

        public VariableDefinition(DataVariableType type) : base(IntPtr.Zero) {
            _onGet = OnGet;
            _onSet = OnSet;
            _onSize = OnSize;
            _onChild = OnChild;

            NativePtr = Native.VariableDefinition.Create(
                type,
                _onGet,
                _onSet,
                _onSize,
                _onChild
            );

            ManuallyRegisterCache(NativePtr, this);
        }

        private bool OnGet(IntPtr ptr, ref IntPtr variantPtr)
        {
            var res = Get(ptr, out var variant);
            variantPtr = variant.NativePtr;
            return res;
        }

        private bool OnSet(IntPtr ptr, IntPtr variantPtr)
        {
            var variant = new Variant(variantPtr);
            return Set(ptr, variant);
        }

        private int OnSize(IntPtr ptr)
        {
            return Size(ptr);
            return 0;
        }

        private IntPtr OnChild(IntPtr ptr, string name, int index)
        {
            return Child(ptr, name, index).NativePtr;
            return IntPtr.Zero;
        }

        public abstract DataVariable Child(IntPtr ptr, string name, int index);
        public abstract int Size(IntPtr ptr);
        public abstract bool Set(IntPtr ptr, Variant variant);
        public abstract bool Get(IntPtr ptr, out Variant variant);
    }

    public enum DataVariableType : int {
        Scalar = 0,
        Array,
        Struct
    }
}
